### ISSUE TITLE
CS8619 False positive

### ISSUE THREAD

**Issue by Logerfo** [dotnet/roslyn#40757]
## Version Used
Visual Studio 2019 (16.5.0 Preview 1.0)

## Steps to Reproduce
```c#
using System.Threading.Tasks;
#nullable enable
public class C {
    public Task<(bool Result, string? Extras)> M() {
        return Task.FromResult((true, ""));
    }
}
```
[SharpLab](https://sharplab.io/#v2:CYLg1APgAgDABFAjAVgNwFgBQBiAdgVwBtCBDAI0IFM5LdyqsoBmBAJjgGE4BvLOfhCygA2ADwAKMgHsphOACVKAZyIAXADQJEMAPxwAogA9VAJxJKAlAD44AWXEWefAS6gB2BMIB0AMRNSAW0UVQlVxcVN8Sk0AIhiLCwxMFwBfLBSgA===)

## Expected Behavior
Since `string` is a valid `string?`, no warning should be presented.

## Actual Behavior
> Warning CS8619: Nullability of reference types in value of type 'Task<(bool, string)>' doesn't match target type 'Task<(bool Result, string? Extras)>'.

**Comment by jcouv** [comment]
This is currently by-design, as the type parameter of `Task` is invariant. That said, LDM discussed this today and we're going to look into adding some special treatment for `Task<T>` and possibly `ValueTask<T>` to treat them as covariant.

Relates to https://github.com/dotnet/roslyn/issues/40759
FYI @cston

**Comment by CyrusNajmabadi** [comment]
> today and we're going to look into adding some special treatment for Task<T> and possibly ValueTask<T> to treat them as covariant.

Woudl be great ot have an attribute to allow other types to play here IMO :)

**Comment by DBJDBJ** [comment]
One year after is there any movement on this one?

**Comment by CyrusNajmabadi** [comment]
@DBJDBJ see linked csharplang issues.

**Comment by inf9144** [comment]
Anything new here? Got my self in this situation, not only with Task<T> but also with MyType<T>,

Have a method and ant make an overload for this because string is a reference type and those types are identical:
void Xyz<TFrom>(Expression<Func<TFrom, MyType<string>>> param);
void Xyz<TFrom>(Expression<Func<TFrom, MyType<string?>>> param);

My hacky workaround would be something like
void Xyz<TFrom>(Expression<Func<TFrom, MyType<TString>>> param)
  where TString: IEnumerable<char>?
  
This should really be fixed because it affects all generic types and their usage as parameters.....
string is fullfilling string? so i cant get why this warning appears on the direction of notnull to nullable assignements. It's a bug in my eye, no matter if we talk about variance here.

**Comment by CyrusNajmabadi** [comment]
>  It's a bug in my eye, no matter if we talk about variance here.

Sorry, variance is very significant.  Just because something is assignable doesn't mean you can replace one with the other.  For example, all strings are objects.  But that doesn't make all lists of strings into lists of objects.  

> Have a method and cant make an overload for this

Correct, and that will always be the case (and is unrelated to this issue).  Those types are identical as far as the runtime is concerned.  So they literally cannot both exist at the same time.

**Comment by inf9144** [comment]
> Correct, and that will always be the case (and is unrelated to this issue). Those types are identical as far as the runtime is concerned. So they literally cannot both exist at the same time.

Exactly and that's why you need the possibility to declare a function which can take T<X?> and does not CS8619 on providing an T<X>. If you dont have something like that - there is now way out of this warning. You dont see this as a problem? It hit's every function with a generic type parameter. If your type does not declare some interface or baseclass and you dont have the luxury to declare your method with T<TX> where TX : X? you always hit CS8619. Or am i missing something?

**Comment by inf9144** [comment]
> > It's a bug in my eye, no matter if we talk about variance here.
> 
> Sorry, variance is very significant. Just because something is assignable doesn't mean you can replace one with the other. For example, all strings are objects. But that doesn't make all lists of strings into lists of objects.
With the difference that we dont have different types here. As long we are talking about reference types `T<X?>` and` T<X>` are completely identical with the only difference that X ist constrained to not being null on one of this signatures. So when my signature allows for T<X?> which range of values is bigger (by the value NULL)  i cant see why this should be an error.

**Comment by CyrusNajmabadi** [comment]
>  So when my signature allows for T<X?> which range of values is bigger (by the value NULL) i cant see why this should be an error.

Because *from teh runtime's perspective*, these are identical.  It literally has no way to distinguish this.

**Comment by CyrusNajmabadi** [comment]
> Exactly and that's why you need the possibility to declare a function which can take T<X?> and does not CS8619 on providing an T<X>

This is a variance issue, as stated several times already.  Maybe we can solve variance for non-interface/non-delegate types.  Perhaps in a generalized way.  Or maybe just for the well known Task types, which we know should behave variantly in reality.

**Comment by inf9144** [comment]
If you see this as variance issue it would be a problem to just declare in or out on the generic type parameters like you can do on interfaces, because this would not be what you would like to express. You dont want the class working/castable/assignable for `T<XBase>` or `T<XDerived>`. This would also come with the cost that every contravariance or covariance in C# brings - you cannot use the generic type in all situations on the generic class/struct anymore. (Return Values, Method Parameters, ...). 
Can you understand why i dont match this problem to type variance but to nullability? If this is a variance problem, you would need to have the ability to explicitly define the variance in means of nullability. So having an invariant type which is nullable co-/contravariant. Only then you could relax those restrictions that normally come with covariance/contravariance in general, because (yes on runtime level) you can be sure you have the very same types (in case of reference types).

**Comment by CyrusNajmabadi** [comment]
> You dont want the class working/castable/assignable for `T<XBase>` or `T<XDerived>`

Why not?

If i have `Task<string>` then it would be sensible to assign that to a `Task<string?>`.  Tasks have natural variance (they're a read-only view of computation after all).

> Can you understand why i dont match this problem to type variance

No.

>  If this is a variance problem, you would need to have the ability to explicitly define the variance in means of nullability.

Variance already works with nullability.  It's just limited to interfaces and delegates.

If we had an `ITask<T>` instead of a `Task<T>`, this would have just worked.

**Comment by CyrusNajmabadi** [comment]
@inf9144 

> You are welcome to take part in this conversation instead of just putting thumbs down emojis on someones post ;-)

People can engage however they want (within the rules) :)

**Comment by inf9144** [comment]
> > Can you understand why i dont match this problem to type variance
> 
> No.

I think i get your point now - sorry it was a long day in germany it's nearly 1 o clock am. If we think about some `ValueHolder<T>` and i would assign it to `ValueHolder<T?>` and call a method `SetValue` this method would accept now a null value which it should not accept. So T must be covariant so this cannot happen and we only have getter that use T. 

I think also i know why you cant get my point, i will try to clarify by example. Let's say you have some method for this struct `ValueHolder<T>` (which is invariant).

```
public struct ValueHolder<T> {
   public T Value { get; set; }
}
```

And you want to do things with T, which you can access by getter, dependent on what T is. Normally you can overload a method so we maybe have an overload for `ValueHolder<int>` and `ValueHolder<int?>`:

 ```
  public static void DoSth(ValueHolder<int?> valueHolder) {

   }

   public static void DoSth(ValueHolder<int> valueHolder) {

   }
```

Now you want to do the same for a reference type. This way it will give you CS8619:
```
   public static void DoSth(ValueHolder<MyClass?> valueHolder) {

   }
```

 Right now you can handle the nullability by just introducing the type as generic method parameter so you come around CS8619. And it behaves correctly because i cant just assign something that does not fit the generic constraint. But i can easily get the value and work with it like it was of type MyClass.

```
  public static void DoSth<TMyClass>(ValueHolder<TMyClass> valueHolder) where TMyClass : MyClass? {
      var value = valueHolder.Value;
      if (value is null) {
         //...
      } else {
         //...
      }
   }
```

But when we come to types like string this is not possible. This is my problem - i ran in the wrong direction and made the wrong points but sadly i cant declare this:

```
   public static void DoSth<TString>(ValueHolder<TString> valueHolder) where TString : string? {
      var value = valueHolder.Value;
      if (value is null) {
         //...
      } else {
         //...
      }
   }
```
It does not work for system types, sealed types and so on. This is my problem. And it's not solvable if i want to keep the invariance on my class. The only way i can make this work is to give those overloads different names which makes my API much harder to use. Or i could introduce interfaces that are covariant - but boxing on structs is advised against afaik, that's why you shouldnt use much interfaces on them.

So basically if we see string as derived type of string? and covariance is the solution, the type handling in C# is wrong because it assumes there are no derived types of string and sealed classes. That's my problem - this is a break in the type system, because nullable reference types are not implemented as real types in the type system and all constraints that you could normally put on such a function wont work. Its assymetric handling, which does not feel right. That's why i stumbled across your variance argument and didnt see it fit.

I hope i could explain what my painpoint is and look happily forward to your comment @CyrusNajmabadi. Thanks in advance! :-)

**Comment by CyrusNajmabadi** [comment]
>  So T must be covariant so this cannot happen and we only have getter that use T.

Correct.  And, as this is a discussion about `Task<>` that could potentially apply.

**Comment by CyrusNajmabadi** [comment]
> Now you want to do the same for a reference type. This way it will give you CS8619:

This is unrelated to this discussion.  Task is not an invariant type.  You might want ot consider opening a different discussion.

Note: as mentioned a few times: there is nothing we can do about this.  This is a *runtime limitation*.  From teh runtime perspective, these are the same types, so you cannot overload that way.

:)

**Comment by inf9144** [comment]
> This is unrelated to this discussion. Task is not an invariant type. You might want ot consider opening a different discussion.
Would like to - is the rest of my post and the problem understandable / comprehensive enough? 
>  From teh runtime perspective, these are the same types, so you cannot overload that way.
I still dont want to have two overloads but instead declare a method with a fitting generic parameter to solve it (which ends up in CS0701). I hoped that my explanation above would show it at the end of the post :-/
I dont have the feeling that i can express myself technically correct on this topic or that i get misunderstood ;-)
Is this a topic for csharplang or roslyn?

**Comment by CyrusNajmabadi** [comment]
> Would like to - is the rest of my post and the problem understandable / comprehensive enough?

Yup :)

> Would like to - is the rest of my post and the problem understandable / comprehensive enough?

Right.  The problem is: there's no way to express the concept you want.  Because there cannot be two overloads, you have to have a single one, which means either disallowing null, or allowing it.  You can't have it both ways.  Known issue, c'est la vie :)

You can def start the discussion.  But without an actual solution that is reasonable to implement, it likely won't go anywhere.  But it can't hurt to try :)

**Comment by ds1709** [comment]
I don't know is it by design or a bug:
```c#
public class C
{
    public Task<(bool Result, string? Extras)> M()
        => Task.FromResult((true, "")); // warning

    public Task<(bool Result, string? Extras)> N()
        => Task.FromResult((true, ""))!; // no warning, adding `!` suppress it
}
```
Visual Studio 17.12.0, C# 12.

**Comment by CyrusNajmabadi** [comment]
> I don't know is it by design or a bug:
> ```c#
> public class C
> {
>     public Task<(bool Result, string? Extras)> M()
>         => Task.FromResult((true, "")); // warning
> 
>     public Task<(bool Result, string? Extras)> N()
>         => Task.FromResult((true, ""))!; // no warning, adding `!` suppress it
> }
> ```
> Visual Studio 17.12.0, C# 12.

It's by design as those have different types.

**Comment by ds1709** [comment]
> It's by design as those have different types.

To be clear, it's about method `N`. In method `M` it's all OK.
But  in method `N` `null-forgeting` must suppress warning if I'm trying to return null as result, but it must not have effect for task generic parameter type... or not?
